src/match_data_object.cpp

```diff
@@ -6,24 +6,36 @@ Value MatchDataObject::array(int start) {
     auto size = (size_t)(m_region->num_regs - start);
     auto array = new ArrayObject { size };
     for (int i = start; i < m_region->num_regs; i++) {
-        if (m_region->beg[i] == -1) {
-            array->push(NilObject::the());
-        } else {
-            const char *str = &m_string->c_str()[m_region->beg[i]];
-            size_t length = m_region->end[i] - m_region->beg[i];
-            array->push(new StringObject { str, length });
-        }
+        array->push(group(i));
     }
     return array;
 }
 
-size_t MatchDataObject::index(size_t index) {
+long MatchDataObject::index(size_t index) {
     if (index >= size()) return -1;
     return m_region->beg[index];
 }
 
-Value MatchDataObject::group(Env *env, size_t index) {
-    if (index >= size()) return NilObject::the();
+long MatchDataObject::ending(size_t index) {
+    if (index >= size()) return -1;
+    return m_region->end[index];
+}
+
+/**
+ * Return the capture indicated by the given index. This function supports
+ * negative indices as well, provided they are within one overall length of the
+ * number of captures.
+ */
+Value MatchDataObject::group(int index) {
+    if (index + m_region->num_regs <= 0 || index >= m_region->num_regs)
```

Comment:
Moved the range checking into this function (as opposed to ref) since I needed to reference it in the slice! implementation.